🧹  Linting

faddat opened the pull request:
## Description

This PR updates the linters so that we no longer use deprecated linters, enforces linting in CI, and fixes linter errors.

closes: #1417 

Please note that this changes a number of important names, but likely most important:

every time we named something with Id, that has been changed to ID per revive.



---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

faddat commented:
The first run failed due to timeout, so I gave it a longer timeout.

crodriguezvega reviewed (changes requested):
Thanks for these improvements, @faddat!

I think it would also make sense to add a brief explanation of the API breaking changes in the migration docs. If you prefer, we can take care of that ourselves in a separate PR.

faddat commented on CHANGELOG.md:
```diff
@@ -41,7 +41,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### API Breaking
-
+* (linting) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) Fix linting errors, resulting in mock.NewMockIBCApp as mock.NewIBCApp and such, and consistency on the reciever for fee as f.
```
oh, sure!

And you're correct about this-- I had wondered how to note these, thanks a bunch for the guidance! ❤️

faddat commented on modules/core/ante/ante.go:
```diff
@@ -8,20 +8,20 @@ import (
 )
 
-type AnteDecorator struct {
+type Decorator struct {
```
ah, I can tell you:

It ends up called as:

ante.AnteDecorator "and that stutters, consider ante.Decorator"

faddat replied:
done

AdityaSripal commented on modules/core/02-client/legacy/v100/store.go:
```diff
@@ -154,7 +153,8 @@ func pruneSolomachineConsensusStates(clientStore sdk.KVStore) {
 // These keys were not included in the previous release of the IBC module. Adding the iteration keys allows
 // for pruning iteration.
-func addConsensusMetadata(ctx sdk.Context, clientStore sdk.KVStore, cdc codec.BinaryCodec, clientState *ibctmtypes.ClientState) error {
+//nolint:unparam // Though error is aways nil, we need to run this so pruning works
```
Why do we need the return argument here? Can't we remove it?

faddat replied:
we're using the `nakedret` linter, which checks for functions that don't explicitly state what they return.  Or, do you think it's better to not have any return on this function?

AdityaSripal replied:
An error is never returned here. So it seems fine to me to just remove. cc: @damiannolan

faddat replied:
OK

faddat replied:
done

AdityaSripal commented on modules/core/04-channel/types/events.go:
```diff
@@ -22,9 +22,9 @@ const (
 	EventTypeTimeoutPacketOnClose = "timeout_on_close_packet"
 
-	// NOTE: DEPRECATED in favor of AttributeKeyDataHex
+	// Deprecated: in favor of AttributeKeyDataHex
```
should Deprecated be in all caps here?

AdityaSripal commented on testing/app.go:
```diff
@@ -27,7 +27,7 @@ import (
 )
 
-var DefaultTestingAppInit func() (TestingApp, map[string]json.RawMessage) = SetupTestingApp
+var DefaultTestingAppInit = SetupTestingApp
```
why was the type declaration deleted here?

faddat replied:
Revive let me know that it was not needed, I think because the type is implied on the other side of =

AdityaSripal reviewed (commented):
Thanks Jacob, I'll defer to the implementation team here. I just have a couple of questions/suggestions

faddat commented:
Note: some of these defy normal convention in Cosmos.  What I would like to do is come up with a standard set of linters that we can use for SDK 46 onwards, so that we end up not triggering liter errors and needing to use golint.  I started work on that over at Osmosis, and eventually hope to apply that work to the Cosmos SDK.

damiannolan commented on modules/core/04-channel/keeper/events.go:
```diff
@@ -127,11 +127,12 @@ func EmitChannelCloseConfirmEvent(ctx sdk.Context, portID string, channelID stri
 
 // EmitSendPacketEvent emits an event with packet data along with other packet information for relayer
 // to pick up and relay to other chain
+
```
Unnecessary line break

```suggestion
// EmitSendPacketEvent emits an event with packet data along with other packet information for relayer
// to pick up and relay to other chain
```

damiannolan commented on modules/core/04-channel/keeper/events.go:
```diff
@@ -127,11 +127,12 @@ func EmitChannelCloseConfirmEvent(ctx sdk.Context, portID string, channelID stri
 		sdk.NewEvent(
 			types.EventTypeSendPacket,
-			sdk.NewAttribute(types.AttributeKeyData, string(packet.GetData())), // DEPRECATED
+			sdk.NewAttribute(types.AttributeKeyData, string(packet.GetData())), //nolint:staticcheck // DEPRECATED
```
Should these deprecation notices not be above and something like:

```
// Deprecated: types.AttributeKeyData is deprecated in favour of types.AttributeKeyDataHex
```

Even still, do we need deprecation comments here if the attribute key itself is marked as deprecated in the `types` package?

faddat replied:
honestly do not know

damiannolan commented on modules/core/04-channel/types/packet.go:
```diff
@@ -129,7 +129,7 @@ func (p PacketId) Validate() error {
-func NewPacketId(portId, channelId string, seq uint64) PacketId {
-	return PacketId{PortId: portId, ChannelId: channelId, Sequence: seq}
+// NewPacketID returns a new instance of PacketId
+func NewPacketID(portID, channelID string, seq uint64) PacketId {
```
I personally have wanted this changed for a while! Thanks 🙏

faddat commented on .golangci.yml:
```diff
@@ -42,12 +40,6 @@ issues:
     - text: "Use of weak random number generator"
       linters:
         - gosec
-    - text: "comment on exported var"
-      linters:
-        - golint
-    - text: "don't use an underscore in package name"
-      linters:
-        - golint
```
Great idea

faddat replied:
I think those were turned off, not on

faddat replied:
fyi, I double checked.  these were exclusions that turned these rules off. removing them ensures they'll be treated as rules.

faddat commented on modules/core/02-client/legacy/v100/genesis_test.go:
```diff
@@ -147,7 +147,7 @@ func (suite *LegacyTestSuite) TestMigrateGenesisSolomachine() {
 	}
 
-	bz, err := clientCtx.JSONCodec.MarshalJSON(&expectedClientGenState)
+	bz, err := clientCtx.Codec.MarshalJSON(&expectedClientGenState)
```
JSONCodec is deprecated in the sdk, hence the change (so, technically it doesn’t need to but probably should)

faddat commented:
I think that the conversation here leans towards getting an idea of what should change in the sdk, and I’ll go have a stab at that, and link this PR to the issue/pr/discussion there.

AdityaSripal commented:
> Maybe something like RedundancyDecorator would be more informative

Agreed, `ante.AnteDecorator` is bad but we should have some description of what this particular Decorator is doing. Especially since we may want to add more antedecorators to the package in the future

faddat commented:
I can make the RedundancyDecorator change, no problem :)

faddat commented on modules/light-clients/07-tendermint/types/misbehaviour_handle.go:
```diff
@@ -50,12 +50,10 @@ func (cs ClientState) CheckMisbehaviourAndUpdateState(
 		// Header1 is at greater height than Header2, therefore Header1 time must be less than or equal to
 		// Header2 time in order to be valid misbehaviour (violation of monotonic time).
-		if tmMisbehaviour.Header1.SignedHeader.Header.Time.After(tmMisbehaviour.Header2.SignedHeader.Header.Time) {
-			return nil, sdkerrors.Wrap(clienttypes.ErrInvalidMisbehaviour, "headers are not at same height and are monotonically increasing")
-		}
+	} else if tmMisbehaviour.Header1.SignedHeader.Header.Time.After(tmMisbehaviour.Header2.SignedHeader.Header.Time) {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidMisbehaviour, "headers are not at same height and are monotonically increasing")
```
I agree with you.

faddat commented:
So, I've added the godoc linter.

Should I also add errcheck?

faddat commented on modules/apps/transfer/simulation/params.go:
```diff
@@ -18,13 +18,13 @@ func ParamChanges(r *rand.Rand) []simtypes.ParamChange {
 			func(r *rand.Rand) string {
 				sendEnabled := RadomEnabled(r)
-				return fmt.Sprintf("%s", types.ModuleCdc.MustMarshalJSON(&gogotypes.BoolValue{Value: sendEnabled}))
+				return fmt.Sprintf("%s", types.ModuleCdc.MustMarshalJSON(&gogotypes.BoolValue{Value: sendEnabled})) //nolint:gosimple
```
it doesn't like the usage of Sprintf, but honestly I couldn't figure out the suggestion it gave.

colin-axner replied:
Is it possible the fmt.Sprintf is unnecessary if it is just formatting a string into a string?

faddat commented:
My commit comment probably should have been that I added error checks, not that I added errors 😉 . I did look into ways to check Go-Doc but I actually couldn't find anything and if anybody can find anything I'm more than happy to add the appropriate linter.

faddat commented:
https://github.com/cosmos/ibc-go/pull/1418/commits/b86edf725c3202891a1f00b79ca41dc601da24c1 <- is a flub.

I am checking out https://github.com/cosmos/ibc-go/pull/1418/commits/72b9199f168345011d6c2e5f1f6120a689e708ed

deleting this branch, and pushing it again.  

b86edf725c3202891a1f00b79ca41dc601da24c1 <- merged in notional-labs/main and I am concerned that it could have caused unintended changes.

faddat commented on modules/core/02-client/abci.go:
```diff
@@ -24,9 +24,10 @@
 			bz := k.MustMarshalConsensusState(upgradedConsState)
 
-			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
+			k.SetUpgradedConsensusState(ctx, plan.Height, bz) //nolint:errcheck // this could make the chain halt if we checked the error.
```
resolved

colin-axner commented on CHANGELOG.md:
```diff
@@ -61,6 +64,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 * (cleanup) [\#1335](https://github.com/cosmos/ibc-go/pull/1335/) `gofumpt -w -l .` to standardize the code layout more strictly than `go fmt ./...`
 * (middleware) [\#1022](https://github.com/cosmos/ibc-go/pull/1022) Add `GetAppVersion` to the ICS4Wrapper interface. This function should be used by IBC applications to obtain their own version since the version set in the channel structure may be wrapped many times by middleware. 
+* (modules/core/04-channel) [\#1160](https://github.com/cosmos/ibc-go/pull/1160) Improve `uint64 -> string` performance in `Logger`.
+* (modules/core/04-channel) [\#1232](https://github.com/cosmos/ibc-go/pull/1232) Updating params on `NewPacketID` and moving to bottom of file. 
+* (modules/core/04-channel) [\#1279](https://github.com/cosmos/ibc-go/pull/1279) Add selected channel version to MsgChanOpenInitResponse and MsgChanOpenTryResponse. Emit channel version during OpenInit/OpenTry
```
were these moved? Maybe a result of merge conflicts?

faddat replied:
oh and they're in 2 places?

faddat replied:
ahhh I see.  I will fix.

colin-axner commented on .github/workflows/golangci.yml:
```diff
@@ -0,0 +1,46 @@
+    tags:
+      - v*
+    branches:
+      - master
```
we don't have a master branch, so this can be removed?

colin-axner commented on .github/workflows/golangci.yml:
```diff
@@ -0,0 +1,46 @@
+        uses: golangci/golangci-lint-action@v3.2.0
+        with:
+          # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
+          version: latest
```
I'm wondering if we should use a version. That way if a new version causes our code to break the linter check, we could fix it in the pr which updates the version of the linter, rather than having the linter break on a random day

faddat replied:
Here's my thinking on this, basically I figure it's actually better if it breaks on a random day and discovers an issue that was previously hidden. If you don't mind with that thinking that's totally fine and I can put a version on it but generally speaking I look at stuff like the linter as best if it is rolling so that we always use the latest.

faddat replied:
Anyhow, just lmk your thoughts, don't really mind either way-- just look at this as a possible security enhancement

robert-zaremba replied:
I second @faddat - if it breaks, then you can fix version and make a lint update in a following PR.

colin-axner replied:
Makes sense to me!

colin-axner commented on modules/core/ante/ante.go:
```diff
@@ -8,20 +8,20 @@ import (
 }
 
-// AnteDecorator returns an error if a multiMsg tx only contains packet messages (Recv, Ack, Timeout) and additional update messages
+// AnteHandle returns an error if a multiMsg tx only contains packet messages (Recv, Ack, Timeout) and additional update messages
```
```suggestion
// RedundancyDecorator returns an error if a multiMsg tx only contains packet messages (Recv, Ack, Timeout) and additional update messages
```

colin-axner commented on modules/core/ante/ante.go:
```diff
@@ -8,20 +8,20 @@ import (
 )
 
-type AnteDecorator struct {
+type RedundancyDecorator struct {
```
Thoughts on `RedundantRelayDecorator`?

damiannolan replied:
I like @colin-axner's suggestion of `RedundantRelayDecorator`. It's probably a little more descriptive :)

colin-axner commented on testing/mock/ibc_app.go:
```diff
@@ -88,8 +88,8 @@ type MockIBCApp struct {
 }
 
 // NewMockIBCApp returns a MockIBCApp. An empty PortID indicates the mock app doesn't bind/claim ports.
```
```suggestion
// NewIBCApp returns a IBCApp. An empty PortID indicates the mock app doesn't bind/claim ports.
```

crodriguezvega replied:
I think this suggestion from @colin-axner is still valid.

colin-axner commented on modules/core/02-client/abci.go:
```diff
@@ -26,7 +26,8 @@ func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
 			bz := k.MustMarshalConsensusState(upgradedConsState)
 
-			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
+			k.SetUpgradedConsensusState(ctx, plan.Height, bz) //nolint:errcheck // this could make the chain halt if we checked the error.
```
```suggestion
			k.SetUpgradedConsensusState(ctx, plan.Height, bz) //nolint:errcheck 
```
Is there a way to silence the code scanning alert? In this situation [`SetUpgradedConsensusState` never returns an error](https://github.com/cosmos/cosmos-sdk/blob/main/x/upgrade/keeper/keeper.go#L229). It is just by mistake I believe that the API includes a returned error

colin-axner commented on modules/light-clients/07-tendermint/types/update.go:
```diff
@@ -128,7 +128,10 @@ func (cs ClientState) CheckHeaderAndUpdateState(
-	IterateConsensusStateAscending(clientStore, pruneCb)
+	err = IterateConsensusStateAscending(clientStore, pruneCb)
+	if err != nil {
+		panic(err)
```
```suggestion
		return nil, nil, err
```

colin-axner reviewed (commented):
Thanks @faddat. Sorry for the delay and thanks for your patience. LGTM, left a few minor comments

faddat commented on modules/core/02-client/abci.go:
```diff
@@ -24,9 +24,10 @@
 			bz := k.MustMarshalConsensusState(upgradedConsState)
 
-			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
+			k.SetUpgradedConsensusState(ctx, plan.Height, bz) //nolint:errcheck 
```
hm.  I really don't know.  Maybe I can feed it a blank?

But-- plz advise:

that's safe?

I do knot **know** or I'd do it more often.

faddat replied:
ok, so I went with:

```go
			err := k.SetUpgradedConsensusState(ctx, plan.Height, bz)
			if err != nil {
				return
			}
```

faddat commented on modules/core/02-client/abci.go:
```diff
@@ -24,9 +24,13 @@
 			bz := k.MustMarshalConsensusState(upgradedConsState)
 
-			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
+			err := k.SetUpgradedConsensusState(ctx, plan.Height, bz)
```
@crodriguezvega sir it seems I truly do not know how to make this warning go away.... I could make SetUpgradedConsensusState not return error, wdyt?  I think I need to follow that up the chain to another method but should be doable.

faddat commented:
@crodriguezvega hey, what if we:

* get this into v4 / main
* put the docs in a separate PR?

I am happy to write that separate PR, but getting this in reduces review scope for v4->v5.

crodriguezvega commented on e2e/testsuite/relayer.go:
```diff
@@ -17,6 +17,7 @@ const (
 )
 
 // newCosmosRelayer returns an instance of the go relayer.
+
```
Why would a new line be added here?

faddat replied:
I do not know.  Can fix it though!

faddat replied:
So, I think I know why and it has to do with the demonic possession of my vscode.

crodriguezvega commented on CHANGELOG.md:
```diff
@@ -43,6 +43,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### API Breaking
 
+* (core/04-channel) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) NewPacketId has been renamed to NewPacketID to comply with go linting rules.
```
nit
```suggestion
* (core/04-channel) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) `NewPacketId` has been renamed to `NewPacketID` to comply with go linting rules.
```

crodriguezvega commented on CHANGELOG.md:
```diff
@@ -43,6 +43,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
+* (core/04-channel) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) NewPacketId has been renamed to NewPacketID to comply with go linting rules.
+* (core/ante) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) AnteDecorator has been renamed to RedundancyDecorator to comply with go linting rules and to give more clarity to the purpose of the Decorator. 
+* (testing) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) MockIBCApp has been renamed to IBCApp and MockEmptyAcknowledgement has been renamed to EmptyAcknowledgement to comply with go linting rules
```
nit
```suggestion
* (testing) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) `MockIBCApp` has been renamed to `IBCApp` and `MockEmptyAcknowledgement` has been renamed to `EmptyAcknowledgement` to comply with go linting rules
```

crodriguezvega commented on CHANGELOG.md:
```diff
@@ -43,6 +43,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 ### API Breaking
 
+* (core/04-channel) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) NewPacketId has been renamed to NewPacketID to comply with go linting rules.
+* (core/ante) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) AnteDecorator has been renamed to RedundancyDecorator to comply with go linting rules and to give more clarity to the purpose of the Decorator. 
```
nit
```suggestion
* (core/ante) [\#1418](https://github.com/cosmos/ibc-go/pull/1418) `AnteDecorator` has been renamed to `RedundancyDecorator` to comply with go linting rules and to give more clarity to the purpose of the Decorator. 
```

crodriguezvega reviewed (commented):
Thank you, @faddat. 

Can you please address @damiannolan's comments [here](https://github.com/cosmos/ibc-go/pull/1418/files#r888402984) and [here](https://github.com/cosmos/ibc-go/pull/1418/files#r888416018)?

crodriguezvega commented:
> @crodriguezvega hey, what if we:
> 
> * get this into v4 / main
> * put the docs in a separate PR?
> 
> I am happy to write that separate PR, but getting this in reduces review scope for v4->v5.

Yes, I think it is fine to merge this PR and then add the migration docs later.

Should we merge the PRs in the following order? 
1. #1418
2. #1734 
3. #1653

faddat commented:
Makes sense to me.  Will just address that merge conflict.

faddat commented:
Does `make goimports` work for you?

faddat commented:
Makegoimports seems to work differently:

* in gitpod
* on my laptop
* in ci (or versus the goimports ci linter)

faddat commented:
@crodriguezvega in fact, the order is probably like:

* #1418 <- contains and enforces gofumpt
* Write migration docs re: the api breaking changes in #1418 
* #1653 <- v5


This branch is fumpted and then that setting is added to .golangci.yml

Another thing I'd love to do round these parts, is document the dev setup, or work to get all that in a container as mentioned in #1748 I think it would help a ton with onboarding.

faddat commented:
Hi, I figured out what's going on with this:

```go
var ack = &types.IncentivizedAcknowledgement{}
```

turning into:

```go
ack := &types.IncentivizedAcknowledgement{}
```


So basically, gofumpt works like gofmt.  It is all or nothing, but more strict.  I looked for ways to mark just that line to not be affected, but could not find one.  So, I think the question becomes:

Do we want to use fumpting as a standard or do we prefer:

declaration > initalization

crodriguezvega commented:
> @crodriguezvega in fact, the order is probably like:
> 
> * [🧹  Linting #1418](https://github.com/cosmos/ibc-go/pull/1418) <- contains and enforces gofumpt
> * Write migration docs re: the api breaking changes in [🧹  Linting #1418](https://github.com/cosmos/ibc-go/pull/1418)
> * [ibc-go v5: cosmos-sdk 46 #1653](https://github.com/cosmos/ibc-go/pull/1653) <- v5
> 
> This branch is fumpted and then that setting is added to .golangci.yml
> 
> Another thing I'd love to do round these parts, is document the dev setup, or work to get all that in a container as mentioned in #1748 I think it would help a ton with onboarding.

We also need to add to the migration docs the API breaking changes from the upgrade to SDK 0.46, so we can create the document as a separate PR after both of your PRs are merged. I think that should be fine.

crodriguezvega commented:
> Hi, I figured out what's going on with this:
> 
> ```go
> var ack = &types.IncentivizedAcknowledgement{}
> ```
> 
> turning into:
> 
> ```go
> ack := &types.IncentivizedAcknowledgement{}
> ```
> 
> So basically, gofumpt works like gofmt. It is all or nothing, but more strict. I looked for ways to mark just that line to not be affected, but could not find one. So, I think the question becomes:
> 
> Do we want to use fumpting as a standard or do we prefer:
> 
> declaration > initalization
> 
> If we enforce gofumpt, it will act like `go fmt` and tell us "it is my way or the highway"
> 
> I kind of like that, but figure that's ultimately the team's choice. Here's why I like it: When I start a new project, I put gofumpt in golang-ci and then it does a great deal of the lifting for me. Also, the project has a pretty strict set of "how stuff should look.
> 
> But based on my quick googling, if gofumpt is enforced, then it is always going to change things like that.
> 
> Right now, that line of code is noncompliant, so gofumpt is failing to pass ci.
> 
> My editor is set to do gofumpt on save, and that's why it was magically popping back to initalization.
> 
> Curious about the reasoning on this, since I think that gofumpt tries to make changes that are 1:1 the same?

Ah, gotcha. Thanks for digging into it and for the explanation. I guess it's better not to try to go against `gofumpt` and we leave it as is. I guess it will be ok as well for @damiannolan, but I will talk to him offline.

crodriguezvega commented:
Maybe a silly question about `gofumpt`: do we need to run it locally on file save so that it formats the file according to its standards before pushing the changes to GitHub?

faddat commented:
I do, but I don't think it is strictly needed.  CI would let you know if it was wrong.

faddat commented:
@crodriguezvega this one is fine to merge first I think.  I will leave both open, and you can decide.

you could use #1734 to make reviewing this easier or if this passes all muster, then just merge this and I’ll close that.

crodriguezvega commented:
Thanks very much for acting quickly on the feedback, @faddat! 

Yes, I also think that this PR should be now ready. Let me discuss with the rest of the team on Monday if they want to do another round of review. In any case I hope we can merge it early next week, and then hopefully soon after also #1653. 

Since #1734 doesn't have many changes, we might as well merge it, but let me see also what the rest of the team thinks.

colin-axner commented on testing/mock/mock.go:
```diff
@@ -4,16 +4,17 @@ import (
 	"encoding/json"
 	"fmt"
 
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+
 	"github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/codec"
 	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
 	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
```
normally we would combine these sets of imports since they are both external

colin-axner replied:
checkout #1770

faddat replied:
on this now

colin-axner replied:
Ah, sorry. I see you fixed it for this file, but I intended to communicate that the imports have been changed in almost all the files in this pr. Is it possible to fix those files as well?

faddat replied:
sure!

faddat replied:
I read the other issue... trying to devise a way to do the desired fix

faddat replied:
https://github.com/incu6us/goimports-reviser

colin-axner reviewed (commented):
Hi @faddat, is it possible to revert the import changes? Was there a reason these were changed? 

Normally we structure imports like so:
```
import (
        // standard library imports
	"fmt"
	"testing"
        
        // external library imports
	"github.com/stretchr/testify/require"
	abci "github.com/tendermint/tendermint/abci/types"
        
         // ibc-go library imports
	"github.com/cosmos/ibc-go/modules/core/23-commitment/types"
)
```

As a side note, we still need to create a best practices document so it is easier to give visibility on our established habits

crodriguezvega commented on e2e/fee_middleware_test.go:
```diff
@@ -2,6 +2,8 @@ package e2e
 
 import (
 	"context"
+	"e2e/testsuite"
```
Sorry to keep bothering you with the imports, but do you mind to put these ones back where they were before?

crodriguezvega commented on e2e/testsuite/testsuite.go:
```diff
@@ -2,13 +2,12 @@ package testsuite
 
 import (
 	"context"
+	"e2e/testconfig"
```
Same comment as above.

crodriguezvega reviewed (commented):
I think there's only 3 more files to fix for the imports, and then it will be over! Thank you very much for your patience and for reordering them back!

crodriguezvega commented on e2e/testsuite/relayer.go:
```diff
@@ -1,9 +1,8 @@
-	"testing"
-
 	"e2e/testconfig"
+	"testing"
```
This one as well.

faddat commented:
I suspect they'll perpetually be grouped with stdlib, but I'll kick that off now.

colin-axner commented on e2e/go.mod:
```diff
@@ -1,4 +1,4 @@
-module e2e
+module github.com/cosmos/ibc-go/e2e
```
would like to have @chatton approval on this change. Is it okay to include it here or should we do it in a separate pr. I guess it doesn't need to be backported so it doesn't matter?

chatton replied:
This change is fine to have here 👍

colin-axner reviewed (commented):
LGTM, only minor concerns left:
* e2e module naming change (think this is likely fine as is)
* RedundancyDecorator renamed to RedundantRelayDecorator? Did we want to do this change?

damiannolan commented:
> LGTM, only minor concerns left:
> 
> * e2e module naming change (think this is likely fine as is)
> * RedundancyDecorator renamed to RedundantRelayDecorator? Did we want to do this change?

Do you think we could additionally add an exception to ignore `.pb.go` files for `goimports` ordering? cc. @faddat 

Thanks for all your work on getting this over the line!

faddat commented:
@vuong177 is re-doing protos now

faddat commented:
RedundancyDecorator renamed to RedundantRelayDecorator? Did we want to do this change?

-- I don't think so?

I'll have a look and try to figure out how it got in because I don't recall making it. maybe it triggered the "stutters" thing?  checking...

colin-axner commented on cmd/build_test_matrix/main.go:
```diff
@@ -50,6 +50,10 @@ func getGithubActionMatrixForTests(e2eRootDirectory string) (GithubActionTestMat
 	testSuiteMapping := map[string][]string{}
 	fset := token.NewFileSet()
 	err := filepath.Walk(e2eRootDirectory, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("error walking e2e directory: %s", err)
+		}
```
looks like formatting is off?

colin-axner replied:
oh nvm, I see the open bracket now

colin-axner commented:
> RedundancyDecorator renamed to RedundantRelayDecorator? Did we want to do this change?

That seemed to be the [conclusion](https://github.com/cosmos/ibc-go/pull/1418#discussion_r920153020) to me, but I don't have a strong preference. It just seems like a more descriptive name. Did it cause an issue with the linter?

colin-axner reviewed (approved):
Great work!! Thanks for following up on all our suggestions. I'm still curious about the Decorator naming, but don't want to block merge on that discussion

crodriguezvega commented:
@faddat There seem to be some conflicts...

faddat commented:
@colin-axner let's keep the name as it is, then.  It doesn't cause an issue with the linter.  In fact, once this is merged, our editors should be able to scream at us about even the most trivial lint issues.  It's 100% golang-ci lint'd.
